bot/cogs/code_snippets.py

```diff
@@ -0,0 +1,216 @@
+    if start_line > len(split_file_contents) or end_line < 1:
+        return ''
+    start_line = max(1, start_line)
+    end_line = min(len(split_file_contents), end_line)
+
+    # Gets the code lines, dedents them, and inserts zero-width spaces to prevent Markdown injection
+    required = '\n'.join(split_file_contents[start_line - 1:end_line])
+    required = textwrap.dedent(required).rstrip().replace('`', '`\u200b')
+
+    # Extracts the code language and checks whether it's a "valid" language
+    language = file_path.split('/')[-1].split('.')[-1]
+    if not language.replace('-', '').replace('+', '').replace('_', '').isalnum():
+        language = ''
+
+    if len(required) != 0:
+        return f'```{language}\n{required}```\n'
+    return ''
+
+
+GITHUB_RE = re.compile(
+    r'https://github\.com/(?P<repo>.+?)/blob/(?P<path>.+/.+)'
+    r'#L(?P<start_line>\d+)([-~]L(?P<end_line>\d+))?\b'
+)
+
+GITHUB_GIST_RE = re.compile(
+    r'https://gist\.github\.com/([^/]+)/(?P<gist_id>[^\W_]+)/*'
+    r'(?P<revision>[^\W_]*)/*#file-(?P<file_path>.+?)'
+    r'-L(?P<start_line>\d+)([-~]L(?P<end_line>\d+))?\b'
+)
+
+GITLAB_RE = re.compile(
+    r'https://gitlab\.com/(?P<repo>.+?)/\-/blob/(?P<path>.+/.+)'
+    r'#L(?P<start_line>\d+)([-](?P<end_line>\d+))?\b'
+)
+
+BITBUCKET_RE = re.compile(
+    r'https://bitbucket\.org/(?P<repo>.+?)/src/(?P<ref>.+?)/'
+    r'(?P<file_path>.+?)#lines-(?P<start_line>\d+)(:(?P<end_line>\d+))?\b'
+)
+
+
+class CodeSnippets(Cog):
+    """
+    Cog that prints out snippets to Discord.
+
+    Matches each message against a regex and prints the contents of all matched snippets.
+    """
+
+    def __init__(self, bot: Bot):
+        """Initializes the cog's bot."""
+        self.bot = bot
+
+    @Cog.listener()
+    async def on_message(self, message: Message) -> None:
+        """Checks if the message has a snippet link, removes the embed, then sends the snippet contents."""
+        gh_match = GITHUB_RE.search(message.content)
+        gh_gist_match = GITHUB_GIST_RE.search(message.content)
+        gl_match = GITLAB_RE.search(message.content)
+        bb_match = BITBUCKET_RE.search(message.content)
+
+        if (gh_match or gh_gist_match or gl_match or bb_match) and not message.author.bot:
+            message_to_send = ''
+
+            for gh in GITHUB_RE.finditer(message.content):
+                message_to_send += await fetch_github_snippet(self.bot.http_session, **gh.groupdict())
+
+            for gh_gist in GITHUB_GIST_RE.finditer(message.content):
+                message_to_send += await fetch_github_gist_snippet(self.bot.http_session, **gh_gist.groupdict())
+
+            for gl in GITLAB_RE.finditer(message.content):
+                message_to_send += await fetch_gitlab_snippet(self.bot.http_session, **gl.groupdict())
+
+            for bb in BITBUCKET_RE.finditer(message.content):
+                message_to_send += await fetch_bitbucket_snippet(self.bot.http_session, **bb.groupdict())
```

Comment:
I feel like there's a DRYer solution available here.

Maybe make a list of tuples, something like this:
```py
pattern_handlers = [
    (GITHUB_RE, fetch_github_snippet),
    (GITHUB_GIST_RE, fetch_github_gist_snippet),
    ...
]
```
And then you can just iterate this like
```py
for pattern, handler in pattern_handlers:
    for match in pattern.finditer(message.content):
        message_to_send += await handler(self.bot.http_session, **match.groupdict())
```

This is a little oversimplified (you still need to do the search), but you get the idea.

Reply:
I think the search is actually unnecessary, so this alone works fine